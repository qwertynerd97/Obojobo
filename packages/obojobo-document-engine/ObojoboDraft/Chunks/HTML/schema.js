const schema = {
	blocks: {
		'ObojoboDraft.Chunks.HTML': {
			nodes: [
				{
					match: [{ object: 'text' }],
					min: 1,
					max: 1
				}
			]
		}
	}
}

export default schema
