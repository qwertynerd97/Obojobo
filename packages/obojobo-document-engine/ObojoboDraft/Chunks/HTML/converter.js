const slateToObo = node => {
	const json = {}
	json.id = node.key
	json.type = node.type
	json.content = {}
	json.content.html = node.text
	json.children = []

	return json
}

const oboToSlate = node => {
	const json = {}
	json.object = 'block'
	json.key = node.id
	json.type = node.type

	json.nodes = [
		{
			object: 'text',
			leaves: [
				{
					text: node.content.html
				}
			]
		}
	]

	return json
}

export default { slateToObo, oboToSlate }
