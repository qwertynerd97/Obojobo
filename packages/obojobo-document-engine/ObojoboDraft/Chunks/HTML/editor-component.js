import React from 'react'

const HTML = props => {
	return (
		<div className={'obojobo-draft--chunks--html viewer pad html-editor'}>{props.children}</div>
	)
}

export default HTML
