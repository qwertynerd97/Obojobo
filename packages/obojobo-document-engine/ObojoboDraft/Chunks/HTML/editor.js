import React from 'react'

const HTML_NODE = 'ObojoboDraft.Chunks.HTML'

const Node = props => {
	return (
		<div className={'component'}>
			<div className={'obojobo-draft--chunks--html viewer pad html-editor'}>
				<pre>{props.children}</pre>
			</div>
		</div>
	)
}

const insertNode = change => {
	change
		.insertBlock({
			type: HTML_NODE
		})
		.moveToStartOfNextText()
		.focus()
}

const slateToObo = node => {
	const json = {}
	json.id = node.key
	json.type = node.type
	json.content = {}
	json.content.html = node.text
	json.children = []

	return json
}

const oboToSlate = node => {
	const json = {}
	json.object = 'block'
	json.key = node.id
	json.type = node.type

	json.nodes = [
		{
			object: 'text',
			leaves: [
				{
					text: node.content.html
				}
			]
		}
	]

	return json
}

const plugins = {
	renderNode(props) {
		switch (props.node.type) {
			case HTML_NODE:
				return <Node {...props} {...props.attributes} />
		}
	},
	onKeyDown(event, change) {
		const isHTML = change.value.blocks.some(block => block.type === HTML_NODE)
		if (!isHTML) return

		// Insert a softbreak on enter
		if (event.key === 'Enter') {
			event.preventDefault()
			return change.insertText('\n')
		}

		// Tab insert
		if (event.key === 'Tab') {
			event.preventDefault()
			change.insertText('\t')
			return true
		}
	},
	schema: {
		blocks: {
			'ObojoboDraft.Chunks.HTML': {
				nodes: [
					{
						match: [{ object: 'text' }],
						min: 1
					}
				]
			}
		}
	}
}

const Heading = {
	components: {
		Node
	},
	helpers: {
		insertNode,
		slateToObo,
		oboToSlate
	},
	plugins
}

export default Heading
