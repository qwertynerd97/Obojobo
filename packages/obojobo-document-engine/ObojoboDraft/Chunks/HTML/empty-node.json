{
	"object":"block",
	"type":"ObojoboDraft.Chunks.HTML",
	"isVoid":false,
	"nodes":[
		{
			"object":"text",
			"leaves":[
				{"object":"leaf","text":"","marks":[]}
			]
		}
	]
}

