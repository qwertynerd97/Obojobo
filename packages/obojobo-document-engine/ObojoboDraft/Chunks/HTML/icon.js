import React from 'react'

const Icon = () => {
	return (
		<svg
			id="Layer_1"
			data-name="Layer 1"
			xmlns="http://www.w3.org/2000/svg"
			viewBox="0 0 16.09 12.79"
		>
			<defs>
				<style
				>{`.cls-1{opacity:0.55;}.cls-2{fill:#231f20;}.cls-3{fill:none;stroke:#000;stroke-linecap:round;stroke-linejoin:round;}`}</style>
			</defs>
			<title>insert-icon</title>
			<g className="cls-1">
				<rect className="cls-2" x="5.55" y="0.59" width="10.54" height="0.84" />
				<rect className="cls-2" x="0.11" y="6.28" width="15.98" height="0.84" />
				<rect className="cls-2" x="0.11" y="11.96" width="15.98" height="0.84" />
				<polyline className="cls-3" points="0.5 0.5 2.98 2.38 0.5 4.27" />
				<rect className="cls-2" x="5.55" y="3.43" width="10.54" height="0.84" />
				<rect className="cls-2" x="0.11" y="9.12" width="15.98" height="0.84" />
			</g>
		</svg>
	)
}

export default Icon
