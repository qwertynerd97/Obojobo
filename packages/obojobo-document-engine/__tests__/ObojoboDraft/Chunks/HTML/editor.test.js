import React from 'react'
import renderer from 'react-test-renderer'

import HTML from '../../../../ObojoboDraft/Chunks/HTML/editor'
const HTML_NODE = 'ObojoboDraft.Chunks.HTML'

describe('HTML editor', () => {
	test('Node builds the expected component', () => {
		const Node = HTML.components.Node
		const component = renderer.create(
			<Node
				node={{
					data: {
						get: () => {
							return {}
						}
					}
				}}
			/>
		)
		const tree = component.toJSON()

		expect(tree).toMatchSnapshot()
	})

	test('insertNode calls change methods', () => {
		const change = {}
		change.insertBlock = jest.fn().mockReturnValueOnce(change)
		change.moveToStartOfNextText = jest.fn().mockReturnValueOnce(change)
		change.focus = jest.fn().mockReturnValueOnce(change)

		HTML.helpers.insertNode(change)

		expect(change.insertBlock).toHaveBeenCalled()
		expect(change.moveToStartOfNextText).toHaveBeenCalled()
		expect(change.focus).toHaveBeenCalled()
	})

	test('slateToObo converts a Slate node to an OboNode with content', () => {
		const slateNode = {
			key: 'mockKey',
			type: 'mockType',
			data: {
				get: () => {
					return {}
				}
			},
			text: 'mockText'
		}
		const oboNode = HTML.helpers.slateToObo(slateNode)

		expect(oboNode).toMatchSnapshot()
	})

	test('oboToSlate converts an OboNode to a Slate node', () => {
		const oboNode = {
			id: 'mockKey',
			type: 'mockType',
			content: { width: 'large' }
		}
		const slateNode = HTML.helpers.oboToSlate(oboNode)

		expect(slateNode).toMatchSnapshot()
	})

	test('oboToSlate converts an OboNode to a Slate node with a caption', () => {
		const oboNode = {
			id: 'mockKey',
			type: 'mockType',
			content: {}
		}
		const slateNode = HTML.helpers.oboToSlate(oboNode)

		expect(slateNode).toMatchSnapshot()
	})

	test('plugins.renderNode renders a button when passed', () => {
		const props = {
			attributes: { dummy: 'dummyData' },
			node: {
				type: HTML_NODE,
				data: {
					get: () => {
						return {}
					}
				}
			}
		}

		expect(HTML.plugins.renderNode(props)).toMatchSnapshot()
	})

	test('plugins.onKeyDown deals with no html', () => {
		const change = {
			value: {
				blocks: [
					{
						type: 'mockType'
					}
				]
			}
		}
		change.insertBlock = jest.fn().mockReturnValueOnce(change)

		const event = {
			key: 'Enter',
			preventDefault: jest.fn()
		}

		HTML.plugins.onKeyDown(event, change)

		expect(event.preventDefault).not.toHaveBeenCalled()
	})

	test('plugins.onKeyDown deals with random keypress', () => {
		const change = {
			value: {
				blocks: [
					{
						type: HTML_NODE
					}
				]
			}
		}
		change.insertBlock = jest.fn().mockReturnValueOnce(change)

		const event = {
			key: 'e',
			preventDefault: jest.fn()
		}

		HTML.plugins.onKeyDown(event, change)

		expect(event.preventDefault).not.toHaveBeenCalled()
	})

	test('plugins.onKeyDown deals with [Enter]', () => {
		const change = {
			value: {
				blocks: [
					{
						type: HTML_NODE
					}
				]
			},
			insertText: jest.fn()
		}

		const event = {
			key: 'Enter',
			preventDefault: jest.fn()
		}

		HTML.plugins.onKeyDown(event, change)
		expect(event.preventDefault).toHaveBeenCalled()
		expect(change.insertText).toHaveBeenCalledWith('\n')
	})

	test('plugins.onKeyDown deals with [Tab]', () => {
		const change = {
			value: {
				blocks: [
					{
						type: HTML_NODE
					}
				]
			},
			insertText: jest.fn()
		}

		const event = {
			key: 'Tab',
			preventDefault: jest.fn()
		}

		HTML.plugins.onKeyDown(event, change)
		expect(event.preventDefault).toHaveBeenCalled()
		expect(change.insertText).toHaveBeenCalledWith('\t')
	})
})
