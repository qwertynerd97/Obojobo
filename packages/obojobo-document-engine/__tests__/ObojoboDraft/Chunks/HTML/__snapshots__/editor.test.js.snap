// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTML editor Node builds the expected component 1`] = `
<div
  className="component"
>
  <div
    className="obojobo-draft--chunks--html viewer pad html-editor"
  >
    <pre />
  </div>
</div>
`;

exports[`HTML editor oboToSlate converts an OboNode to a Slate node 1`] = `
Object {
  "key": "mockKey",
  "nodes": Array [
    Object {
      "leaves": Array [
        Object {
          "text": undefined,
        },
      ],
      "object": "text",
    },
  ],
  "object": "block",
  "type": "mockType",
}
`;

exports[`HTML editor oboToSlate converts an OboNode to a Slate node with a caption 1`] = `
Object {
  "key": "mockKey",
  "nodes": Array [
    Object {
      "leaves": Array [
        Object {
          "text": undefined,
        },
      ],
      "object": "text",
    },
  ],
  "object": "block",
  "type": "mockType",
}
`;

exports[`HTML editor plugins.renderNode renders a button when passed 1`] = `
<Node
  attributes={
    Object {
      "dummy": "dummyData",
    }
  }
  dummy="dummyData"
  node={
    Object {
      "data": Object {
        "get": [Function],
      },
      "type": "ObojoboDraft.Chunks.HTML",
    }
  }
/>
`;

exports[`HTML editor slateToObo converts a Slate node to an OboNode with content 1`] = `
Object {
  "children": Array [],
  "content": Object {
    "html": "mockText",
  },
  "id": "mockKey",
  "type": "mockType",
}
`;
